src/js/modules/utilities/IOBuf.ts

```diff
@@ -206,12 +216,33 @@ export class IOBuf {
     return iob;
   }
 
+  /**
+   * find the first fragment with free space, otherwise it returns the last
+   * fragment in the list.
+   */
   private getLastFragment = (): Fragment => {
-    const lastFragment = this.fragmentList[this.fragmentList.length - 1];
+    const lastFragment =
+      this.fragmentList.find((f) => f.used < f.getSize()) ||
+      this.fragmentList[this.fragmentList.length - 1];
     if (lastFragment) {
       return lastFragment;
     } else {
-      return this.appendNewFragment();
+      return this.appendNewFragment(lastFragment?.getSize());
+    }
+  };
+
+  /**
+   * given a fragment, it tries to find the next fragment in the list , if there
+   * isn't next one, it appends new one.
+   * @param fragment
+   */
+  private getNextFragmentOrAppend = (fragment: Fragment): Fragment => {
+    const currentFragment = this.fragmentList.findIndex((f) => f === fragment);
```

Comment:
Just a question, how long could you expect the fragment list to possibly be? If it could be frequently long, maybe it might be a good idea to maintain a fragment index cache

Reply:
If the common case if that this list is never that long, maybe that optimization would be overly complex for little gain

Reply:
I change it, I create a new index for it